[Tan De Shao] iP

tandeshao opened the pull request:
# Duke #
> "Why am i doing this, why does the code not work." - Some sad single student out in soc 

**Features include:**
- level-1
- level-2
- ~~level-3~~ **Level-4 NEW AND IMPROVED!**

**All you need to do is**
1. download it from [here](https://www.youtube.com/watch?v=jl9xKw_xEas&ab_channel=TopTunes).
2. Double click it.
3. add your tasks
4. let it manage your tasks for you 😃 

and it is **FREE**!

blahblah:
- [x] black sheep ?
- [x] and the hoo?
- [x] 3 black wolves?
- [x] chicken rice?
***
if you programmer, you can use it to practice java too. Here's the ``main`` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

tandeshao committed:
Made a simple skeleton code for duke

tandeshao committed:
level-2 code

tandeshao committed:
level-3 code

tandeshao committed:
Level-4 code

tandeshao committed:
Level-4 code

tandeshao committed:
Level-4 code

tandeshao committed:
testing

tandeshao committed:
add exceptions

tandeshao committed:
delete added

tandeshao committed:
new exception added

tandeshao committed:
word doc added

tandeshao committed:
Merge branch 'feature1'

tandeshao committed:
delete word docs

tandeshao committed:
level 7 changes

tandeshao committed:
level-7 changes

tandeshao committed:
level-8 changes

tandeshao committed:
Merge tag 'Level-7'

tandeshao committed:
changed

tandeshao committed:
more oop

tandeshao committed:
created packages

tandeshao committed:
tests added

tandeshao committed:
created jar file

tandeshao committed:
javadocs added

tandeshao committed:
coding style added

tandeshao committed:
find implemented

tandeshao committed:
Merge tag 'Level-9'

tandeshao committed:
conflict resolved

tandeshao committed:
Merge tag 'A-CodingStandard'

kerbysoh commented on src/main/java/DukeComponent/Ui.java:
```diff
@@ -0,0 +1,58 @@
+        while (sc.hasNext()) {
+            String userInput = sc.nextLine();
+            try {
+
```
There's an unnecessary blank line here.

kerbysoh commented on src/main/java/DukeComponent/Command.java:
```diff
@@ -0,0 +1,131 @@
+        int start = userInput.indexOf(" ") + 1;
+        int end = userInput.lastIndexOf('/');
+        String task = end == -1 ? "" : userInput.substring(start, end - 1);
+        String details = split.length > 1 ? split[1].substring(3) : "";
```
Avoid the use of magic numbers here like 1 and 3

kerbysoh commented on src/main/java/DukeComponent/Command.java:
```diff
@@ -0,0 +1,131 @@
+            return;
+        }
+
+        switch (action) {
```
You can make use of the default branch here to handle your commands that are invalid

kerbysoh commented on src/main/java/Tasks/Tasks.java:
```diff
@@ -0,0 +1,79 @@
+abstract public class Tasks {
+
+    private String task;
+    private Boolean marked;
```
Boolean here should be named isMarked rather than marked

kerbysoh commented on src/main/java/Tasks/Tasks.java:
```diff
@@ -0,0 +1,79 @@
+     * @param s String representation of a date.
+     * @return Date that follows the format if applicable.
+     */
+    public static String returnDate(String s) {
```
Function name could be better named to describe the function (To convert date from one form to another). May I suggest using convertDate instead?

kerbysoh reviewed (commented):
Well done! Code is very well documented, clean and well written. There are just a few pointers I've noticed in my review that you may want to look into. :)

yusufaine commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,52 @@
+    /**
+     * {@link Ui}
+     */
+    private final Ui ui;
```
solid use of `@link` 👍

yusufaine commented on src/main/java/DukeComponent/Command.java:
```diff
@@ -0,0 +1,131 @@
+        System.out.println(tasks.remove(i));
+        String s = String.format("Now you have %d tasks in the list.", tasks.size());
+        System.out.println(s);
+    }
```
You can consider moving all your print statements into `UI` so that your code is less cluttered and easier to focus on what the logic is (even though it's quite short here, but still)

yusufaine commented on src/main/java/DukeComponent/Parser.java:
```diff
@@ -0,0 +1,78 @@
+                && !command.equals("delete") && !command.equals("mark") && !command.equals("unmark")
+                && !command.equals("todo") && !command.equals("deadline") && !command.equals("event") &&
+                !command.equals("find"));
+    }
```
This can be replaced by adding a default case as mentioned by @\kerbysoh in `Command`

yusufaine commented on src/main/java/DukeComponent/TaskList.java:
```diff
@@ -0,0 +1,98 @@
+     * @param description  Description of the Task.
+     */
+    public void find(String description) {
+        int count = 1;
```
I think this variable a bit redundant since you're not using/returning it.

```suggestion

```

yusufaine commented on src/main/java/DukeComponent/TaskList.java:
```diff
@@ -0,0 +1,98 @@
+        for (Tasks t : ls) {
+            if (t.getTask().contains(description)) {
+                System.out.println(count + "." + t);
+                count++;
```
```suggestion

```

yusufaine commented on src/main/java/DukeComponent/TaskList.java:
```diff
@@ -0,0 +1,98 @@
+            }
+        };
+
+    }
```
I did this too for mine initially but then I realised that you don't actually have to implement an iterator and create a getter for your taskList:

```suggestion
    public ArrayList<Tasks> getList() {
        return this.ls;
    }
```

that way when you want to do a for-each, you can call:
* `taskList.getList().forEach(...)` or 
* `for (Task t : taskList.getList())`

yusufaine commented on src/main/java/DukeComponent/Ui.java:
```diff
@@ -0,0 +1,58 @@
+package DukeComponent;
```
Mentioned earlier but you can consider moving all your printline stuff here since this class is meant to interact with the user.

yusufaine commented on src/main/java/Tasks/Tasks.java:
```diff
@@ -0,0 +1,79 @@
+ * A class that belongs to the Tasks Package.
+ * This class encapsulates the logic of how Tasks should be represented in Duke.
+ */
+abstract public class Tasks {
```
nit-picky: I think this class should be named `Task` instead

yusufaine commented on src/main/java/Tasks/DeadLines.java:
```diff
@@ -0,0 +1,43 @@
+     * Message to be displayed for DeadLine as a Task in Duke.
+     * @return String representation of DeadLine.
+     */
+    @Override
+    public String toString() {
+        if (this.getMarked()) {
+            return "[D]" + "[X" + "] " + this.getTask() + " (by: " + returnDate(this.date) + ")";
+        } else {
+            return "[D]" + "[ " + "] " + this.getTask() + " (by: " + returnDate(this.date) + ")";
+        }
+    }
```
You can consider using `String.format()` to make this a little more readable, here's an example of mine (although slightly different use-case):

```java
public String toFile() {
    String markStatus = super.isMarked ? "1" : "0";
    SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy HH:mm");
    return String.format("%s | %s | %s | %s\n", "D", markStatus, super.taskName, sdf.format(this.by));
}
```

yusufaine reviewed (commented):
Overall clean code, not much to comment about apart from some nit-picky pointers.

tandeshao committed:
Merge branch 'master' into add-gradle-support

tandeshao committed:
gradle can run

tandeshao committed:
added some changes

tandeshao committed:
update build file

tandeshao committed:
UI working

tandeshao committed:
testing of gradle completed by adding srcset

tandeshao committed:
Merge branch 'add-gradle-support'

tandeshao committed:
some changes

tandeshao committed:
gui implemented for duke

tandeshao committed:
done

fantablack commented on src/main/java/component/TaskList.java:
```diff
@@ -0,0 +1,100 @@
+     * @return An iterator for Tasks.
+     */
+    @Override
+    public Iterator<Tasks> iterator() {
+        return new Iterator<>() {
+            private int currentIndex = 0;
+            private final int currentSize = ls.size();
+
+            @Override
+            public boolean hasNext() {
+                return currentIndex < currentSize && ls.get(currentIndex) != null;
+            }
+
+            @Override
+            public Tasks next() {
+                return ls.get(currentIndex++);
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException();
+            }
+        };
```
I like this use of an Iterator to iterate through the tasks compared to the very common for/foreach loops! 👍

fantablack commented on src/main/java/exceptions/DeadlineException.java:
```diff
@@ -0,0 +1,25 @@
+ * This class encapsulates the message that should be displayed when users parsed an invalid input
+ * for the {@link tasks.DeadLines} class.
+ */
+public class DeadlineException extends TaskException {
```
I like that you used inheritance to differentiate the different types of exceptions, instead of simply using different Strings in the respective exceptions' messages

fantablack reviewed (commented):
Overall, I liked that your code is clean and easy to follow. Interesting use of iterator to process your Lists. Keep up the good work!

tandeshao committed:
create jar file with shadowJar

tandeshao committed:
Change README file to show a more accurate account of the project

Previous version of README is the template README file provided to us.
Subsequent iterations of the project has already added more features and
the original README file no longer accurately describes the project.

tandeshao committed:
Assert path statements in MainWindow.class()

tandeshao committed:
Change files such that they adhere to coding-standards specified by cs2103 convention.

Methods in the staged files are modified so that they follow the SLAP principle.

tandeshao committed:
add help commmand

tandeshao committed:
Merge tag 'BCD-Extension'

tandeshao committed:
Merge branch 'branch-A-Assertions'

tandeshao committed:
Add comment

tandeshao committed:
Merge pull request #1 from tandeshao/branch-A-CodeQuality

Comments added

tandeshao committed:
add help

tandeshao committed:
Merge pull request #2 from tandeshao/BCD-Extension

add help

tandeshao committed:
add assertions

tandeshao committed:
Merge pull request #3 from tandeshao/branch-A-Assertions

add assertions

tandeshao committed:
deleted

tandeshao committed:
Change picture for duke and user

tandeshao committed:
Set theme jekyll-theme-slate

tandeshao committed:
Improve Ui

tandeshao committed:
Merge branch 'master' of https://github.com/tandeshao/ip

tandeshao committed:
Add Readme

tandeshao committed:
Add comments

tandeshao committed:
Delete Ui.PNG

tandeshao committed:
Final touches

tandeshao committed:
Merge branch 'master' of https://github.com/tandeshao/ip

tandeshao committed:
Add Ui.png

tandeshao committed:
Change comments to fit convention

tandeshao committed:
Change name of app

tandeshao committed:
Update Ui

tandeshao committed:
Improve user guide

tandeshao committed:
Improve user guide v0.1

tandeshao committed:
Improve user guide v0.2

tandeshao committed:
Improve user guide v0.2

tandeshao committed:
Improve user guide v0.2

tandeshao committed:
Improve user guide v0.3

tandeshao committed:
Improve user guide v0.4

tandeshao committed:
Improve user guide v0.5

tandeshao committed:
Improve user guide v0.6

tandeshao committed:
Improve user guide v0.7

tandeshao committed:
Improve user guide v0.8

tandeshao committed:
Improve user guide v0.8

tandeshao committed:
Set theme jekyll-theme-midnight

tandeshao committed:
Set theme jekyll-theme-midnight

tandeshao committed:
Improve user guide v0.9

tandeshao committed:
Merge branch 'master' of https://github.com/tandeshao/ip
